input_definition.go

```diff
@@ -209,19 +197,34 @@ type Action struct {
 	RowID            *uint64           `json:"rowID,omitempty"`
 }
 
-// Encode converts Action into its internal representation.
-func (o *Action) Encode() *internal.InputDefinitionAction {
-	// TODO: this check needs to happen somewhere other than Encode()
-	/*
-		if o.RowID == nil && o.ValueDestination == InputSingleRowBool {
-			return nil, errors.New("rowID required for single-row-boolean")
+// Validate ensures the input definition action conforms to our specification.
+func (a *Action) Validate() error {
+	if a.Frame == "" {
+		return ErrFrameRequired
+	}
+	validValues := make(map[string]bool)
```

Comment:
it may be easier to read code that uses a function like
```
func stringInSlice(a string, list []string) bool {
    for _, b := range list {
        if b == a {
            return true
        }
    }
    return false
}
```
instead of using a map like `validValues` just for comparison.

that way, you just need:
```
if !stringInSlice(a.ValueDestination, validValueDestination) {
	return fmt.Errorf(...)
}
```